File: SuperiorPackGroup/Reports/ProductionXtraReport.vb
Imports System.Linq
Imports DevExpress
Imports DXDAL.SPGData

Public Class ProductionXtraReport

    Private m_Minutes As Long
    Private m_BreakMinutes As Long
    Private m_TotalMinutes As Long = 0
    Private m_Expected As Single
    Private m_TotalExpected As Single = 0
    Private m_TotalDifference As Single = 0
    Private m_TotalActualSpeed As Single = 0
    Private m_TotalPayroll As Double = 0
    Private m_CurrentProduction As Production
    Private m_MinPerHour As Integer
    Private m_MachineStandard As Double
    Private m_MachinePackers As Double

    Private Sub totalHoursDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalHoursDetailXrLabel.BeforePrint

        If totalHoursDetailXrLabel.Visible Then
            totalHoursDetailXrLabel.Text = String.Format("{0}:{1}", Format((m_Minutes - m_BreakMinutes) \ 60, "0"), Format(((m_Minutes - m_BreakMinutes) Mod 60) \ 1, "00"))
        End If

    End Sub

    Private Sub totalMinutesDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalMinutesDetailXrLabel.BeforePrint

        If totalMinutesDetailXrLabel.Visible Then
            totalMinutesDetailXrLabel.Text = Format(((m_Minutes - m_BreakMinutes) / 60) * m_MinPerHour, "#.##")
        End If

    End Sub

    Private Sub expectedDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles expectedDetailXrLabel.BeforePrint

        If expectedDetailXrLabel.Visible Then
            '= Convert.ToSingle(((m_Minutes - m_BreakMinutes) / 60)  * Convert.ToSingle(Me.GetCurrentColumnValue("ItemProdStandard")))
            If m_MachineStandard = 0 Then
                m_Expected = Convert.ToSingle(GetCurrentColumnValue("ItemProdStandard")) * CInt((m_Minutes - m_BreakMinutes) / 60 * m_MinPerHour) / CInt(GetCurrentColumnValue("intQtyPerUnit"))
            Else
                m_Expected = CSng(m_MachineStandard * CInt((m_Minutes - m_BreakMinutes) / 60 * m_MinPerHour) / CInt(GetCurrentColumnValue("intQtyPerUnit")))
            End If 

                expectedDetailXrLabel.Text = Format(m_Expected, "#.##")
            m_TotalExpected += m_Expected
        End If

    End Sub

    Private Sub differenceDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles differenceDetailXrLabel.BeforePrint

        If differenceDetailXrLabel.Visible Then
            Dim difference As Single = Convert.ToSingle(Convert.ToSingle(Me.GetCurrentColumnValue("ProdMainQuantity")) - m_Expected)

            differenceDetailXrLabel.Text = Format(difference, "#.##")
            m_TotalDifference += difference
        End If

    End Sub

    Private Sub actualSpeedDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles actualSpeedDetailXrLabel.BeforePrint

        If actualSpeedDetailXrLabel.Visible Then
            Dim actualSpeed As Single

            actualSpeed = ((Convert.ToSingle(Me.GetCurrentColumnValue("ProdMainQuantity")) * Convert.ToSingle(Me.GetCurrentColumnValue("intQtyPerUnit"))) / _
                           CType(((m_Minutes - m_BreakMinutes) / 60) * m_MinPerHour, Integer))
            If Single.IsInfinity(actualSpeed) OrElse Single.IsNaN(actualSpeed) Then

            Else
                actualSpeedDetailXrLabel.Text = Format(actualSpeed, "#.##")
                m_TotalActualSpeed += actualSpeed
            End If
        End If

    End Sub

    Private Sub Detail_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles Detail.BeforePrint

        m_Minutes = DateDiff(DateInterval.Minute, CType(Me.GetCurrentColumnValue("ProdMainTimeStart"), DateTime), CType(Me.GetCurrentColumnValue("ProdMainTimeStop"), DateTime))
        m_BreakMinutes = BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), _
                                                      Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")))
        m_TotalMinutes += m_Minutes - m_BreakMinutes

        m_CurrentProduction = Xpo.XpoDefault.Session.GetObjectByKey(Of Production)(CInt(GetCurrentColumnValue("ProdMainID")))

        Dim MachineStandard As ItemMachineStandards
        If m_CurrentProduction IsNot Nothing Then
            MachineStandard = m_CurrentProduction.ProdMainItemID.MachineStandards.Where(Function(IMS) IMS.Machine.MachineLineID = m_CurrentProduction.ProdMainMachineLine.MachineLineID).FirstOrDefault()
            If MachineStandard IsNot Nothing Then
                If MachineStandard.MinutesPerShift > 0 Then
                    m_MinPerHour = CInt(MachineStandard.MinutesPerShift / 7.16667)
                Else If m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0
                    m_MinPerHour = CInt(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667)
                Else
                    m_MinPerHour = MainXtraForm.MinPerHOur
                End If
                If MachineStandard.ProductionPerMinute > 0 Then
                    m_MachineStandard = MachineStandard.ProductionPerMinute
                Else
                    m_MachineStandard = 0
                End If
                If MachineStandard.ProjectedPackers > 0 Then
                    m_MachinePackers = MachineStandard.ProjectedPackers
                Else
                    m_MachinePackers = 0
                End If
            Else
                If m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0
                    m_MinPerHour = CInt(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667)
                Else
                    m_MinPerHour = MainXtraForm.MinPerHOur
                End If
                m_MachineStandard = 0
                m_MachinePackers = 0
            End If
        Else
            m_MinPerHour = MainXtraForm.MinPerHOur
            m_MachineStandard = 0
            m_MachinePackers = 0
        End If

    End Sub

    Private Sub totalPayrollDetailXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalPayrollDetailXrLabel.BeforePrint

        If totalPayrollDetailXrLabel.Visible Then
            Dim totalPayroll As Double = Convert.ToDouble(GetCurrentColumnValue("ProdMainPackers")) * (m_Minutes - m_BreakMinutes) / 60 * MainXtraForm.PackersPay
            Me.totalPayrollDetailXrLabel.Text = Format(totalPayroll, "$0.00")
            m_TotalPayroll += totalPayroll
        End If

    End Sub

    Private Sub totalHoursFooterXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalHoursFooterXrLabel.BeforePrint

        If totalHoursFooterXrLabel.Visible Then
            totalHoursFooterXrLabel.Text = String.Format("{0}:{1}", Format(m_TotalMinutes \ 60, "#0"), Format((m_TotalMinutes Mod 60) \ 1, "00"))
        End If

    End Sub

    Private Sub expectedFooterXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles expectedFooterXrLabel.BeforePrint

        If expectedFooterXrLabel.Visible Then
            expectedFooterXrLabel.Text = Format(m_TotalExpected, "#,#")
        End If

    End Sub

    Private Sub differenceFooterXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles differenceFooterXrLabel.BeforePrint

        If differenceFooterXrLabel.Visible Then
            differenceFooterXrLabel.Text = Format(m_TotalDifference, "#,#")
        End If

    End Sub

    Private Sub actualSpeedFooterXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles actualSpeedFooterXrLabel.BeforePrint

        If actualSpeedFooterXrLabel.Visible Then
            actualSpeedFooterXrLabel.Text = Format(m_TotalActualSpeed / m_TotalMinutes, "0.00")
        End If

    End Sub

    Private Sub totalPayrollFooterXrLabel_BeforePrint(ByVal sender As System.Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles totalPayrollFooterXrLabel.BeforePrint

        If totalPayrollFooterXrLabel.Visible Then
            totalPayrollFooterXrLabel.Text = Format(m_TotalPayroll, "$#,##0.00")
        End If

    End Sub

    Private Sub ReportHeader_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs) Handles ReportHeader.BeforePrint

        m_Minutes = 0
        m_TotalMinutes = 0
        m_TotalExpected = 0
        m_TotalDifference = 0
        m_TotalActualSpeed = 0
        m_TotalPayroll = 0

    End Sub

    Private Sub uomDetailXrLabel_BeforePrint(sender As Object, e As Drawing.Printing.PrintEventArgs) Handles uomDetailXrLabel.BeforePrint

        If m_CurrentProduction IsNot Nothing Then
            uomDetailXrLabel.Text = m_CurrentProduction.ProdMainItemID.strUnitOfMeasure
        End If

    End Sub

    Private Sub machineSpeedDetailXrLabel_BeforePrint(sender As Object, e As Drawing.Printing.PrintEventArgs) Handles machineSpeedDetailXrLabel.BeforePrint

        If m_MachineStandard = 0 Then
            If m_CurrentProduction IsNot Nothing Then
                machineSpeedDetailXrLabel.Text = CType(m_CurrentProduction.ProdMainItemID.ItemProdStandard, String)
            End If
        Else
            machineSpeedDetailXrLabel.Text = m_MachineStandard.ToString()
        End If

    End Sub

    Private Sub projectedPackersDetailXrLabel_BeforePrint(sender As Object, e As Drawing.Printing.PrintEventArgs) Handles projectedPackersDetailXrLabel.BeforePrint

        If m_MachinePackers = 0 Then
            If m_CurrentProduction IsNot Nothing Then
                projectedPackersDetailXrLabel.Text = CType(m_CurrentProduction.ProdMainItemID.Packers, String)
            End If
        Else
            projectedPackersDetailXrLabel.Text = m_MachinePackers.ToString()
        End If

    End Sub

    Private Sub reasonCodeDetailXrLabel_BeforePrint(sender As Object, e As Drawing.Printing.PrintEventArgs) Handles reasonCodeDetailXrLabel.BeforePrint

        If m_CurrentProduction IsNot Nothing Then
            reasonCodeDetailXrLabel.Text = m_CurrentProduction.Reasons
        End If

    End Sub

End Class

File: SuperiorPackGroup/Reports/ProductionXtraReport.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using System.Linq;
using DevExpress;
using DXDAL.SPGData;

namespace SuperiorPackGroup
{
	public partial class ProductionXtraReport
	{

		public ProductionXtraReport()
		{
			InitializeComponent();
		}

		private long m_Minutes;
		private long m_BreakMinutes;
		private long m_TotalMinutes = 0;
		private float m_Expected;
		private float m_TotalExpected = 0;
		private float m_TotalDifference = 0;
		private float m_TotalActualSpeed = 0;
		private double m_TotalPayroll = 0;
		private Production m_CurrentProduction;
		private int m_MinPerHour;
		private double m_MachineStandard;
		private double m_MachinePackers;

		private void totalHoursDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (totalHoursDetailXrLabel.Visible)
			{
				totalHoursDetailXrLabel.Text = string.Format("{0}:{1}", ((m_Minutes - m_BreakMinutes) / 60).ToString("0"), (((m_Minutes - m_BreakMinutes) % 60) / 1).ToString("00"));
			}

		}

		private void totalMinutesDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (totalMinutesDetailXrLabel.Visible)
			{
				totalMinutesDetailXrLabel.Text = (((m_Minutes - m_BreakMinutes) / 60.0) * m_MinPerHour).ToString("#.##");
			}

		}

		private void expectedDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (expectedDetailXrLabel.Visible)
			{
				//= Convert.ToSingle(((m_Minutes - m_BreakMinutes) / 60)  * Convert.ToSingle(Me.GetCurrentColumnValue("ItemProdStandard")))
				if (m_MachineStandard == 0)
				{
					m_Expected = Convert.ToSingle(GetCurrentColumnValue("ItemProdStandard")) * Convert.ToInt32((m_Minutes - m_BreakMinutes) / (double)60 * m_MinPerHour) / Convert.ToInt32(GetCurrentColumnValue("intQtyPerUnit"));
				}
				else
				{
					m_Expected = (float)(m_MachineStandard * Convert.ToInt32((m_Minutes - m_BreakMinutes) / (double)60 * m_MinPerHour) / Convert.ToInt32(GetCurrentColumnValue("intQtyPerUnit")));
				}

					expectedDetailXrLabel.Text = m_Expected.ToString("#.##");
				m_TotalExpected += m_Expected;
			}

		}

		private void differenceDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (differenceDetailXrLabel.Visible)
			{
				float difference = Convert.ToSingle(Convert.ToSingle(this.GetCurrentColumnValue("ProdMainQuantity")) - m_Expected);

				differenceDetailXrLabel.Text = difference.ToString("#.##");
				m_TotalDifference += difference;
			}

		}

		private void actualSpeedDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (actualSpeedDetailXrLabel.Visible)
			{
				float actualSpeed = ((Convert.ToSingle(this.GetCurrentColumnValue("ProdMainQuantity")) * Convert.ToSingle(this.GetCurrentColumnValue("intQtyPerUnit"))) / Convert.ToInt32(((m_Minutes - m_BreakMinutes) / 60.0) * m_MinPerHour));

				if (float.IsInfinity(actualSpeed) || float.IsNaN(actualSpeed))
				{

				}
				else
				{
					actualSpeedDetailXrLabel.Text = actualSpeed.ToString("#.##");
					m_TotalActualSpeed += actualSpeed;
				}
			}

		}

		private void Detail_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			m_Minutes = DateHelper.DateDiff(DateHelper.DateInterval.Minute, Convert.ToDateTime(this.GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(this.GetCurrentColumnValue("ProdMainTimeStop")));
			m_BreakMinutes = BreakTimeBLL.GetBreakMinutes(Convert.ToDateTime(GetCurrentColumnValue("ProdMainDate")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStart")), Convert.ToDateTime(GetCurrentColumnValue("ProdMainTimeStop")));
			m_TotalMinutes += m_Minutes - m_BreakMinutes;

			m_CurrentProduction = DevExpress.Xpo.XpoDefault.Session.GetObjectByKey<Production>(Convert.ToInt32(GetCurrentColumnValue("ProdMainID")));

			ItemMachineStandards MachineStandard = null;
			if (m_CurrentProduction != null)
			{
				MachineStandard = m_CurrentProduction.ProdMainItemID.MachineStandards.Where((IMS) => IMS.Machine.MachineLineID == m_CurrentProduction.ProdMainMachineLine.MachineLineID).FirstOrDefault();
				if (MachineStandard != null)
				{
					if (MachineStandard.MinutesPerShift > 0)
					{
						m_MinPerHour = Convert.ToInt32(MachineStandard.MinutesPerShift / 7.16667);
					}
					else if (m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0)
					{
						m_MinPerHour = Convert.ToInt32(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667);
					}
					else
					{
						m_MinPerHour =  MainXtraForm.MinPerHOur;
					}
					if (MachineStandard.ProductionPerMinute > 0)
					{
						m_MachineStandard = MachineStandard.ProductionPerMinute;
					}
					else
					{
						m_MachineStandard = 0;
					}
					if (MachineStandard.ProjectedPackers > 0)
					{
						m_MachinePackers = MachineStandard.ProjectedPackers;
					}
					else
					{
						m_MachinePackers = 0;
					}
				}
				else
				{
					if (m_CurrentProduction.ProdMainItemID.MinutesPerShift > 0)
					{
						m_MinPerHour = Convert.ToInt32(m_CurrentProduction.ProdMainItemID.MinutesPerShift / 7.1667);
					}
					else
					{
						m_MinPerHour =  MainXtraForm.MinPerHOur;
					}
					m_MachineStandard = 0;
					m_MachinePackers = 0;
				}
			}
			else
			{
				m_MinPerHour =  MainXtraForm.MinPerHOur;
				m_MachineStandard = 0;
				m_MachinePackers = 0;
			}

		}

		private void totalPayrollDetailXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (totalPayrollDetailXrLabel.Visible)
			{
				double totalPayroll = Convert.ToDouble(GetCurrentColumnValue("ProdMainPackers")) * (m_Minutes - m_BreakMinutes) / 60 *  MainXtraForm.PackersPay;
				this.totalPayrollDetailXrLabel.Text = totalPayroll.ToString("$0.00");
				m_TotalPayroll += totalPayroll;
			}

		}

		private void totalHoursFooterXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (totalHoursFooterXrLabel.Visible)
			{
				totalHoursFooterXrLabel.Text = string.Format("{0}:{1}", m_TotalMinutes / 60, (m_TotalMinutes % 60) / 1);
			}

		}

		private void expectedFooterXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (expectedFooterXrLabel.Visible)
			{
				expectedFooterXrLabel.Text = m_TotalExpected.ToString("#,#");
			}

		}

		private void differenceFooterXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (differenceFooterXrLabel.Visible)
			{
				differenceFooterXrLabel.Text = m_TotalDifference.ToString("#,#");
			}

		}

		private void actualSpeedFooterXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (actualSpeedFooterXrLabel.Visible)
			{
				actualSpeedFooterXrLabel.Text = (m_TotalActualSpeed / m_TotalMinutes).ToString("0.00");
			}

		}

		private void totalPayrollFooterXrLabel_BeforePrint(System.Object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (totalPayrollFooterXrLabel.Visible)
			{
				totalPayrollFooterXrLabel.Text = m_TotalPayroll.ToString("$#,##0.00");
			}

		}

		private void ReportHeader_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			m_Minutes = 0;
			m_TotalMinutes = 0;
			m_TotalExpected = 0;
			m_TotalDifference = 0;
			m_TotalActualSpeed = 0;
			m_TotalPayroll = 0;

		}

		private void uomDetailXrLabel_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (m_CurrentProduction != null)
			{
				uomDetailXrLabel.Text = m_CurrentProduction.ProdMainItemID.strUnitOfMeasure;
			}

		}

		private void machineSpeedDetailXrLabel_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (m_MachineStandard == 0)
			{
				if (m_CurrentProduction != null)
				{
					machineSpeedDetailXrLabel.Text = Convert.ToString(m_CurrentProduction.ProdMainItemID.ItemProdStandard);
				}
			}
			else
			{
				machineSpeedDetailXrLabel.Text = m_MachineStandard.ToString();
			}

		}

		private void projectedPackersDetailXrLabel_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (m_MachinePackers == 0)
			{
				if (m_CurrentProduction != null)
				{
					projectedPackersDetailXrLabel.Text = Convert.ToString(m_CurrentProduction.ProdMainItemID.Packers);
				}
			}
			else
			{
				projectedPackersDetailXrLabel.Text = m_MachinePackers.ToString();
			}

		}

		private void reasonCodeDetailXrLabel_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e)
		{

			if (m_CurrentProduction != null)
			{
				reasonCodeDetailXrLabel.Text = m_CurrentProduction.Reasons;
			}

		}

	}
}